RIOT-OS/RIOT
# ieee802154/submac: reimplement using FSM

### Contribution description
This PR re-implements the IEEE802.15.4 SubMAC using a Mealy FSM. There are many advantages of using this method:
1. The SubMAC is more robust to hicups such as race condition or slow Bottom Half Processor (e.g SPI radios that process their internal events using the same thread as the SubMAC)
2. It's way easier to debug the SubMAC since the transitions are well known.
3. It's also easier to expand the functionalities of the SubMAC.
4. Automated testing should be way easier as well.
5. Last but not least, this helps to point out implementation issues in radios (this PR piggyback a fix for the CC2538)

Given this PR, the SubMAC seems to be way more stable and tolerant to radio quirks:
E.g:
```
2021-08-13 13:40:24,533 # 
2021-08-13 13:40:24,536 # --- fe80::204:2519:1801:bddb PING statistics ---
2021-08-13 13:40:24,539 # 1000 packets transmitted, 1000 packets received, 0% packet loss
2021-08-13 13:40:24,541 # round-trip min/avg/max = 134.533/145.050/169.359 ms
```
```
2021-08-13 21:06:24,808 # --- fe80::204:2519:1801:bddb PING statistics ---
2021-08-13 21:06:24,814 # 50000 packets transmitted, 49954 packets received, 0% packet loss
2021-08-13 21:06:24,819 # round-trip min/avg/max = 131.011/154.245/216.047 ms
```

Current master:
```
2021-08-12 10:27:16,869 # --- fe80::204:2519:1801:bddb PING statistics ---
2021-08-12 10:27:16,873 # 100 packets transmitted, 100 packets received, 0% packet loss
2021-08-12 10:27:16,876 # round-trip min/avg/max = 135.781/146.318/162.909 ms
```

This PR:
```
2021-08-12 10:26:14,348 # --- fe80::204:2519:1801:bddb PING statistics ---
2021-08-12 10:26:14,355 # 100 packets transmitted, 91 packets received, 9% packet loss
2021-08-12 10:26:14,357 # round-trip min/avg/max = 130.403/143.866/152.149 ms
```

This PR also changes the Radio HAL with a variant of the proposal described [here](https://github.com/RIOT-OS/RIOT/pull/13943#issuecomment-895276277). The RX_LEAVE mechanism was replaced by the RX Continuous cap.



### Testing procedure
Try the following procedures using 2 SubMAC radios (e.g nrf802154 or cc2538)
- Regular ping (12 bytes, 1 second interval)
- Ping with many fragments and short interval (`-s 1024 -c 1000 -i 200)
- Same as before, but with a really short interval (`-s 1024, -c 1000, -i 20)
None of this tests should freeze the SubMAC nor create pktbuf leaks.




### Issues/PRs references
~Depends on #16745~
